Mathlib/CategoryTheory/Limits/Shapes/Pullback/Pasting.lean

```diff
@@ -39,200 +39,389 @@ namespace CategoryTheory.Limits
+  PullbackCone.isLimitOfFlip
+    (leftSquareIsPullback _ hi₂ (PullbackCone.flipIsLimit H₁) (PullbackCone.flipIsLimit H₂))
+
+end PastePullbackVert
+
+section PastePushoutHoriz
+
+/- Let's consider the following diagram
 ```
 X₁ - f₁ -> X₂ - f₂ -> X₃
 |          |          |
 i₁         i₂         i₃
-↓          ↓          ↓
+∨          ∨          ∨
 Y₁ - g₁ -> Y₂ - g₂ -> Y₃
 ```
+where `t₁` denotes the left pushout cocone, and `t₂` denotes the right pushout cocone.
+-/
+variable {X₁ X₂ X₃ Y₁ : C} {f₁ : X₁ ⟶ X₂} {f₂ : X₂ ⟶ X₃} {i₁ : X₁ ⟶ Y₁}
+variable (t₁ : PushoutCocone i₁ f₁) {i₂ : X₂ ⟶ t₁.pt} (t₂ : PushoutCocone i₂ f₂)
+variable (hi₂ : i₂ = t₁.inr)
+
+local notation "Y₂" => t₁.pt
+local notation "g₁" => t₁.inl
+local notation "i₂" => t₁.inr
+local notation "Y₃" => t₂.pt
+local notation "g₂" => t₂.inl
+local notation "i₃" => t₂.inr
+
+/-- The pushout cocone obtained by pasting two pushout cocones horizontally. -/
+abbrev PushoutCocone.pasteHoriz : PushoutCocone i₁ (f₁ ≫ f₂) :=
+  PushoutCocone.mk (g₁ ≫ g₂) i₃
+    (by rw [reassoc_of% t₁.condition, Category.assoc, ← t₂.condition, ← hi₂])
+
+variable {t₁} {t₂}
+
+/-- Given
+```
+X₁ - f₁ -> X₂ - f₂ -> X₃
+|          |          |
+i₁         i₂         i₃
+∨          ∨          ∨
+Y₁ - g₁ -> Y₂ - g₂ -> Y₃
+```
+Then the big square is a pushout if both the small squares are.
+-/
+def pasteHorizIsPushout (H : IsColimit t₁) (H' : IsColimit t₂) :
+    IsColimit (t₁.pasteHoriz t₂ hi₂) := by
+  apply PushoutCocone.isColimitAux'
+  intro s
+  -- Obtain the induced map from descending from both the small squares consecutively.
+  obtain ⟨l₁, hl₁, hl₁'⟩ := PushoutCocone.IsColimit.desc' H s.inl (f₂ ≫ s.inr)
+    (by rw [s.condition, Category.assoc])
+  obtain ⟨l₂, hl₂, hl₂'⟩ := PushoutCocone.IsColimit.desc' H' l₁ s.inr (by rw [← hl₁', hi₂])
+  refine ⟨l₂, by simp [hl₂, hl₁], hl₂', ?_⟩
+  -- Uniqueness also follows from the universal property of both the small squares.
+  intro m hm₁ hm₂
+  apply PushoutCocone.IsColimit.hom_ext H' _ (by simpa [hl₂'] using hm₂)
+  simp only [PushoutCocone.mk_pt, PushoutCocone.mk_ι_app, Category.assoc] at hm₁ hm₂
+  apply PushoutCocone.IsColimit.hom_ext H
+  · rw [hm₁, ← hl₁, hl₂]
+  · rw [← hi₂, reassoc_of% t₂.condition, reassoc_of% t₂.condition, hm₂, hl₂']
+
+variable (t₂)
+
+/-- Given
+
+X₁ - f₁ -> X₂ - f₂ -> X₃
+|          |          |
+i₁         i₂         i₃
+∨          ∨          ∨
+Y₁ - g₁ -> Y₂ - g₂ -> Y₃
 
 Then the right square is a pushout if the left square and the big square are.
 -/
-def rightSquareIsPushout (H : IsColimit (PushoutCocone.mk _ _ h₁))
-    (H' :
-      IsColimit
-        (PushoutCocone.mk _ _
-          (show i₁ ≫ g₁ ≫ g₂ = (f₁ ≫ f₂) ≫ i₃ by
-            rw [← Category.assoc, h₁, Category.assoc, h₂, Category.assoc]))) :
-    IsColimit (PushoutCocone.mk _ _ h₂) := by
-  fapply PushoutCocone.isColimitAux'
+def rightSquareIsPushout (H : IsColimit t₁) (H' : IsColimit (t₁.pasteHoriz t₂ hi₂)) :
```

Comment:
Can you now combine the two in an iff lemma please?
Idem for the vertical case.

Reply:
You can use an equiv instead of an iff. There should be an `ext` lemma that makes this not much more work than if it were an `iff`.